pkg/util/deeptest/statictesthelper.go

```diff
@@ -0,0 +1,262 @@
+		}
+	default:
+		require.Fail(t, "unsupported type", path)
+	}
+}
+
+func (h *staticTestHelper) assertDeepClonedEqual(t require.TestingT, valA, valB reflect.Value, path string) {
+	if h.shouldIgnore(path) {
+		return
+	}
+	if valA.IsValid() != valB.IsValid() {
+		require.Fail(t, "one of them is invalid value", path)
+	}
+	require.Equal(t, valA.Type(), valB.Type(), path+" should have the same type")
+
+	// This function assumes that `a` and `b` are the same type
+	switch valA.Type().Kind() {
+	case reflect.Struct:
+		for i := 0; i < valA.NumField(); i++ {
+			h.assertDeepClonedEqual(t, valA.Field(i), valB.Field(i), path+"."+valA.Type().Field(i).Name)
+		}
+	case reflect.Ptr:
+		if valA.IsNil() && valB.IsNil() {
+			return
+		}
+		// both of them are not nil
+		require.NotEqual(t, 0, valA.Pointer(), path+" should not be nil")
+		require.NotEqual(t, 0, valB.Pointer(), path+" should not be nil")
+
+		if h.shouldComparePointer(path) {
+			require.Equal(t, valA.Pointer(), valB.Pointer(), path+" should be the same")
+		} else {
+			require.NotEqual(t, valA.Pointer(), valB.Pointer(), path+" should be different")
+			h.assertDeepClonedEqual(t, valA.Elem(), valB.Elem(), path)
+		}
+	case reflect.Slice:
+		if h.shouldComparePointer(path) {
+			require.Equal(t, valA.Pointer(), valB.Pointer(), path+" should be the same")
+			require.Equal(t, valA.Len(), valB.Len(), path+" should have the same length")
+		} else {
+			require.NotEqual(t, valA.Pointer(), valB.Pointer(), path+" should not be the same")
+			require.Equal(t, valA.Len(), valB.Len(), path+" should have the same length")
+			for i := 0; i < valA.Len(); i++ {
+				h.assertDeepClonedEqual(t, valA.Index(i), valB.Index(i), path+fmt.Sprintf("[%d]", i))
+			}
+		}
+	case reflect.Array:
+		require.Equal(t, valA.Len(), valB.Len(), path+" should have the same length")
+		for i := 0; i < valA.Len(); i++ {
+			h.assertDeepClonedEqual(t, valA.Index(i), valB.Index(i), path+fmt.Sprintf("[%d]", i))
+		}
+	case reflect.Bool:
+		require.Equal(t, valA.Bool(), valB.Bool(), path+" should be the same")
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		require.Equal(t, valA.Int(), valB.Int(), path+" should be the same")
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		require.Equal(t, valA.Uint(), valB.Uint(), path+" should be the same")
+	case reflect.Float32, reflect.Float64:
+		require.Equal(t, valA.Float(), valB.Float(), path+" should be the same")
+	case reflect.String:
+		require.Equal(t, valA.String(), valB.String(), path+" should be the same")
+	case reflect.Map:
+		if h.shouldComparePointer(path) {
+			require.Equal(t, valA.Pointer(), valB.Pointer(), path+" should be the same")
+		} else {
+			require.NotEqual(t, valA.Pointer(), valB.Pointer(), path+" should not be the same")
+			require.Equal(t, valA.Len(), valB.Len(), path+" should have the same length")
```

Comment:
move this line outside of this if-else block?

Reply:
Done.